### PR TITLE
coreutils: head command

### PR DESCRIPTION
**Pull request by MaheshBharadwaj**
#162: Add head command 
Head command displays the first ten lines of the FILE by default

## Ubuntu output
Screenshot from 2020-10-21 17-28-45

## Implementation output
Screenshot from 2020-10-21 17-34-15

**NOTE:**
- ~~Reading input from STDIN is broken at the moment~~
- Negative count values yet to be supported

### CODE REVIEW

**Comment by sleepy-monax** [comment]
Why is reading from stdin broken?

**Comment by MaheshBharadwaj** [comment]
> Why is reading from stdin broken?

It`ll wait for 1024 characters and then finally echo it

on ubuntu, head echoes each line and echos and stops when interrupted

**Comment by sleepy-monax** [comment]
You can disable buffering with `stream_set_write_buffer_mode` and `stream_set_read_buffer_mode` ;)

**Comment by MaheshBharadwaj** [comment]
> You can disable buffering with `stream_set_write_buffer_mode` and `stream_set_read_buffer_mode` ;)

Will look into it 
maybe ill echo one line and quit?

**Comment by sleepy-monax** [comment]
The echoing is done by the terminal AFAIK

**Comment by MaheshBharadwaj** [comment]
> The echoing is done by the terminal AFAIK

Yes, I will echo once more so that it`ll appear twice, head seems to do the same

**Comment by sleepy-monax** [comment]
> Yes, I will echo once more so that it`ll appear twice, head seems to do the same.
One of the two lines is printed by the terminal itself, head should only print lines once.

**Comment by sleepy-monax** [comment]
I don't really know why you need a StringBuilder. You should be able to do everything on the stack using stream_read/stream_write ;)

**Comment by MaheshBharadwaj** [comment]
> I don't really know why you need a StringBuilder. You should be able to do everything on the stack using stream_read/stream_write ;)

My bad carried over from tac, anything else that needs working?
I will output onto stream using `stream_read` as an when i read, that wouldnt require `StringBuilder` or `String`

**Comment by sleepy-monax** [comment]
I don't think you need to null terminate the buffer :thinking:

File: applications/coreutils/head.cpp [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        {
+            bytes_read = stream_read(input_stream, buffer, BUFFER_SIZE);
+            if (bytes_read == 0)
+            {
+                break;
+            }
+            buffer[bytes_read] = 0;
+
+            size_t buffer_iterator = 0;
+            for (; buffer_iterator < bytes_read; buffer_iterator++)
```
Comment (caelwithcats) [inline applications/coreutils/head.cpp:190]: A while loop would be more suitable IMO
```c++
size_t buffer_iterator = 0;

while(buffer_iterator < bytes_read)
{
    // ...
    buffer_iterator++;
}
```

**Review by caelwithcats — CHANGES_REQUESTED** [review]
Looks good overall, but there are a few things I would suggest

File: applications/coreutils/head.cpp (hunk 1 above)
Comment (MaheshBharadwaj) [reply applications/coreutils/head.cpp:190]: alright

File: applications/coreutils/head.cpp [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+    count_lines = false;
+}
+
+static const char *usages[] = {
+    "[OPTION]... NAME...",
+    nullptr,
+};
+
+static CommandLineOption options[] = {
```
[the remark this answers is not in the document]
Comment (MaheshBharadwaj) [reply applications/coreutils/head.cpp:34]: ah okay

**Comment by MaheshBharadwaj** [comment]
> I don't think you need to null terminate the buffer

wasnt sure if stream_write() stopped at index byte_read or byte_read - 1 that's why i padded with null

**Comment by MaheshBharadwaj** [comment]
Sorry my bad

**Comment by MaheshBharadwaj** [comment]
Willl make this changes

**Comment by sleepy-monax** [comment]
> wasnt sure if stream_write() stopped at index byte_read or byte_read - 1 that's why i padded with null

Next time ask on the discord ;)

File: applications/coreutils/head.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaheshBharadwaj) [reply applications/coreutils/head.cpp:34]: should all the options come before static variables or after them

File: applications/coreutils/head.cpp [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+#define BUFFER_SIZE 1024
+
+static bool count_lines = true;
+static int count = 10;
+static bool quiet = false;
+static bool verbose = false;
+static bool zero_delimiter = false;
+static char delimiter = '\n';
+
+void set_delimiter_zero(CommandLine *cmdline, CommandLineOption *opt)
```
Comment (MaheshBharadwaj) [inline applications/coreutils/head.cpp:13]: this one @caelwithcats

Comment (MaheshBharadwaj) [reply applications/coreutils/head.cpp:13]: needs to be declared for options, or do i give the signature on top and declare it after `main`

Comment (caelwithcats) [reply applications/coreutils/head.cpp:13]: what do you mean?

Comment (caelwithcats) [reply applications/coreutils/head.cpp:13]: yes

**Comment by MaheshBharadwaj** [comment]
@caelwithcats check now, made another commit

**Comment by MaheshBharadwaj** [comment]
@sleepy-monax any changes needed?

**Comment by sleepy-monax** [comment]
Good work, LGTML :)

**Comment by MaheshBharadwaj** [comment]
thanks and sorry for the mention :D

**Pull request merged**